File: python_pkg/python_pkg/complete_node.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.msg import String

class PythonCompleteNode(Node):

    def __init__(self):
        super().__init__("py_node")
        self.get_logger().info("Hello ROS 2")
        self.counter = 0
        
        # Publisher
        self.publisher = self.create_publisher(
            String, "/string", 10
        )
        
        # Subscriber
        self.subscriber = self.create_subscription(
            String, "/string", self.subscriber_callback,10
        )
        
        # Timer
        self.create_timer(0.5, self.publish_string)
        self.create_timer(1.0, self.timer_callback)

        self.get_logger().info("Complete Python Node started!")
        
    def publish_string(self):
        msg = String()
        msg.data = "String message!"
        self.publisher.publish(msg)

    def subscriber_callback(self, message):
        self.get_logger().info("Received msg: " + str(message.data))


    def timer_callback(self):
        self.counter += 1
        self.get_logger().info("Hello: " + str(self.counter))



def main(args=None):
    rclpy.init(args=args)
    node = PythonCompleteNode()
    rclpy.spin(node)
    rclpy.shutdown()

File: cpp_pkg/src/complete_node.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/string.hpp"

class CompleteCppNode:  public rclcpp::Node 
{

    using StringMsg = example_interfaces::msg::String;

public:
    CompleteCppNode(): Node("class_cpp_node"){
        RCLCPP_INFO(this->get_logger(),"Hello C++ Node");
        
        timer = this->create_wall_timer(std::chrono::seconds(1),
            std::bind(&CompleteCppNode::timerCallback, this)
        );

        publishTimer = this->create_wall_timer(std::chrono::seconds(1),
            std::bind(&CompleteCppNode::publishString, this)
        );

        publisher = this->create_publisher<StringMsg>("/sendString",10);

        using namespace std::placeholders;
        subscriber = this->create_subscription<StringMsg>(
            "/receiveString",10, 
            std::bind(&CompleteCppNode::subscriberString, this, _1)
        );

    }
    ~CompleteCppNode(){};

private:

    void publishString(){
        auto msg = StringMsg();
        msg.data = "String message!";
        publisher->publish(msg);
    }

    void subscriberString(const StringMsg::SharedPtr msg){
        RCLCPP_INFO_STREAM(this->get_logger(),
            "Data received: " << msg->data
        );
    }

    void timerCallback(){
        counter++;
        RCLCPP_INFO_STREAM(this->get_logger(), "Hello: " << counter);
    }

    rclcpp::TimerBase::SharedPtr timer;
    rclcpp::TimerBase::SharedPtr publishTimer;
    rclcpp::Publisher<StringMsg>::SharedPtr publisher;
    rclcpp::Subscription<StringMsg>::SharedPtr subscriber;
    int counter = 0;

};



int main(int argc, char **argv){
    rclcpp::init(argc, argv);
    auto node = std::make_shared<CompleteCppNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}